nixpkgs/toolchains/cc.nix

```diff
@@ -0,0 +1,285 @@
+      #
+      # The logic checks whether the flag causes an error message that contains
+      # the flag (or a pattern) verbatim. The assumption is that this will be a
+      # message of the kind `unknown argument: XYZ`. This logic is copied and
+      # adapted to bash from `@bazel_tools//tools/cpp:unix_cc_configure.bzl`.
+      is_compiler_option_supported() {
+        local option="$1"
+        local pattern="''${2-$1}"
+        { $cc "$option" -o /dev/null -c -x c++ - <<<"int main() {}" 2>&1 1>/dev/null || true; } \
+          | grep -qe "$pattern" && return 1 || return 0
+      }
+      is_linker_option_supported() {
+        local option="$1"
+        local pattern="''${2-$1}"
+        { $cc "$option" -o /dev/null -x c++ - <<<"int main() {}" 2>&1 1>/dev/null || true; } \
+          | grep -qe "$pattern" && return 1 || return 0
+      }
+      add_compiler_option_if_supported() {
+        if is_compiler_option_supported "$@"; then
+          echo "$1"
+        fi
+      }
+      add_linker_option_if_supported() {
+        if is_linker_option_supported "$@"; then
+          echo "$1"
+        fi
+      }
+
+      # Determine default include directories.
+      #
+      # This is copied and adapted to bash from
+      # `@bazel_tools//tools/cpp:unix_cc_configure.bzl`.
+      IFS=$'\n'
+      include_dirs_for() {
+        $cc -E -x "$1" - -v "''${@:2}" 2>&1 \
+          | sed -e '1,/^#include <...>/d;/^[^ ]/,$d;s/^ *//' -e 's: (framework directory)::g' \
+          | tr '\n' '\0' \
+          | xargs -0 realpath -ms
+      }
+      CXX_BUILTIN_INCLUDE_DIRECTORIES=($({
+        include_dirs_for c
+        include_dirs_for c++
+        if is_compiler_option_supported -fno-canonical-system-headers; then
+          include_dirs_for c -fno-canonical-system-headers
+          include_dirs_for c++ -std=c++0x -fno-canonical-system-headers
+        elif is_compiler_option_supported -no-canonical-prefixes; then
+          include_dirs_for c -no-canonical-prefixes
+          include_dirs_for c++ -std=c++0x -no-canonical-prefixes
+        fi
+      } 2>&1 | sort -u))
+      unset IFS
+
+      # Determine list of supported compiler and linker flags.
+      #
+      # This is copied and adapted to bash from
+      # `@bazel_tools//tools/cpp:unix_cc_configure.bzl`.
+      COMPILER_FLAGS=(
+        # Security hardening requires optimization.
+        # We need to undef it as some distributions now have it enabled by default.
+        -U_FORTIFY_SOURCE
+        -fstack-protector
+        # All warnings are enabled. Maybe enable -Werror as well?
+        -Wall
+        $(
+          # Enable a few more warnings that aren't part of -Wall.
+          add_compiler_option_if_supported -Wthread-safety
+          add_compiler_option_if_supported -Wself-assign
+          # Disable problematic warnings.
+          add_compiler_option_if_supported -Wunused-but-set-parameter
+          add_compiler_option_if_supported -Wno-unused-command-line-argument
```

Comment:
It made sense on my system to keep this flag for linux, so I ended up putting this line optionally for Darwin. It would be great if that's configurable through the nixpkgs_cc_configure_hermetic.

Reply:
Fair enough. A simple option is to configure the flag by default in `cc.nix` for MacOS and not for Linux. User's can change this configuration by pointing to a different Nix expression using the `nix_file` or `nix_file_content` attribute.